### PR TITLE
DNS: allow propagating DNS responses with no records back to caller (#20890)

### PR DESCRIPTION
**Pull request by wanlill**
Currently ares resolver treats `ARES_ENODATA` and `ARES_ENOTFOUND` as failures and won't propagate the results back to the caller, this makes callers like `StrictDnsClusterImpl` not able to get a chance to detect the change when all records for a name are gone.

Signed-off-by: Wanli Li <wanlil@netflix.com>

Commit Message: DNS: allow propagating DNS responses with no records back to caller (#20890)
Additional Description: Currently ares resolver treats `ARES_ENODATA` and `ARES_ENOTFOUND` as failures and won't propagate the results back to the caller, this makes callers like strict_dns cluster not able to get a chance to detect the change when all records for a name are gone.
Risk Level: High, I did realized that some buggy resolver/recursor might return nxdomain or nodata on failure and this may result in callers like `StrictDnsClusterImpl` removing all the instances. `LogicalDnsCluster` and `RedisCluster` already have checks against success response with no records but we'll change the DNS behavior that can be observed by callers which don't have this check (e.g. `StrictDnsClusterImpl`, in a good way, this is exactly what initiated this change), some of these callers might don't want this behavior change.
Testing: New unit tests added to dns_impl_test.cc
Docs Changes: current.rst changed, added to minor changes section.
Release Notes:
Platform Specific Features:
Runtime guard: `envoy.reloadable_features.cares_accept_nodata` controls individual dns resolution request
Fixes: #20890

### CODE REVIEW

File: api/envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.proto
Comment (mattklein123) [inline api/envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.proto:44]: IMO this is the way it should work by default and this is a bug, however, this is a high risk change, and I think we should runtime guard it (feature flag). WDYT?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for working on this. An API comment to get started.

/wait

File: api/envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.proto
Comment (wanlill) [reply api/envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.proto:44]: +1, this is a high risk change.

I wanted to add a runtime guard but then decided to make it a configurable thing by changing the proto and make it default off because:

1. If this breaks someone then it probably implies their resolver/recursive might not behave correctly and it is probably a long term thing to fix the dns infra, meanwhile they will need to have a way to turn it off.
2. IMO runtime guard is a temp thing and we'll flip it on and remove the guard at some time.

That said, I'm totally fine to add a guard and **make it effective only if both the guard and the config are turned on**, do we want that or just a runtime guard w/o the config option (i.e. change the default behavior but guard it with a flag)?

Comment (mattklein123) [reply api/envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.proto]: I think I would change the default behavior, clearly document it, and add the runtime flag for this behavior change. We can add a config option layer if someone complains about it because they have to flip the flag. APIs are forever and I would rather not add a new API unless we really need to. Thank you!

/wait

Comment (wanlill) [reply api/envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.proto]: SGTM, thanks for the clarification! Can you tell me a little bit more about the "config option layer" you mentioned above? I'm afraid this is something I'm not familiar with (unless it means the proto change we've already been doing here :grinning: )

Comment (mattklein123) [reply api/envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.proto]: Sorry s/layer/later :)

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (yanavlasov) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:384]: Please move this into the constructor initializer list.

File: source/extensions/network/dns_resolver/cares/dns_impl.h
Comment (yanavlasov) [inline source/extensions/network/dns_resolver/cares/dns_impl.h:140]: Please mark it `const` and provide default initializer.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (mattklein123) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:246]: Why is this change and the one below needed?

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (mattklein123) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:141]: nit: you can merge this if into the one above.

File: source/extensions/network/dns_resolver/cares/dns_impl.h
Comment (mattklein123) [inline source/extensions/network/dns_resolver/cares/dns_impl.h:144]: initializer not needed as you initialize in the constructor.

File: test/extensions/network/dns_resolver/cares/dns_impl_test.cc
Comment (mattklein123) [inline test/extensions/network/dns_resolver/cares/dns_impl_test.cc:668]: I don't think we need to parameterize every test. Can you just fix all the tests for the new behavior and then add a specific test with a reverted scoped runtime to make sure the revert behavior works also?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks generally LGTM with small comments.

/wait

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (wanlill) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:246]: They are not needed, I just found it's a bit confusing while debugging it since the lookup family_ has changed for the second request, happy to revert this.

Comment (mattklein123) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:246]: OK got it, makes sense. Fine to keep.

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (yanjunxiang-google) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:32]: This can be a method in DnsResolverImpl::AddrInfoPendingResolution.

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (wanlill) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc]: My bad, thanks for checking the error message! It's somehow added by format fix or vscode :sweat_smile: , fixed.

**Review by mattklein123 — APPROVED** [review]
Thanks this generally LGTM. You will need a main merge. I will defer to @yanavlasov for final approve/merge.

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (yanjunxiang-google) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:218]: Do we need a !dual_resolution check here?  It looks to me it works even without this check, right?

Comment (wanlill) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:218]: I wanted to make sure if AUTO / V4_PREFERRED is used then we'll only accept nodata response for the second/last attempt (i.e. if the first request returns on records, I still want to give the second request targeting a different record type a chance), please let me know if I'm understanding it incorrectly.

Comment (yanjunxiang-google) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:218]: For dual_resolution case, even you set the pending_response_.status_ = ResolutionStatus::Success here, it still go ahead send the 2nd request for a different type:  https://github.com/envoyproxy/envoy/blob/a198e966f15ca1ca81258de6762bb1205c4e7a23/source/extensions/network/dns_resolver/cares/dns_impl.cc#L235.

I guess the real questions is that : if the 1st request returns ARES_ENODATA, and the 2nd request returns a wrong status, like ARES_ECONNREFUSED, which status you want to set it for pending_response_.status_?  IIUC, it should be ResolutionStatus::Success since that's a "SUCCESS" and the 2nd request failure can be ignored in this case.     However, if you add the !dual_resolution check here, the code will not set the pending_response_.status_  to ResolutionStatus::Success in such situation.

Comment (wanlill) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:218]: Good point, I think you're right, I originally put a `completed_ = true` in this block and added this check to make sure it won't block the second request, then I realized that another block already set the completed_ for second request (line #165) so I removed it from this block but somehow got the incorrect impression that entering this block will suppress the second request. Your proposal is more aligned with the how the code is handling `ARES_SUCCESS` (i.e. as long as one request returns `ARES_SUCCESS` the final result is a "SUCCESS"), thanks for pointing it out!

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (yanjunxiang-google) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:130]: I feel it should be better to have the run time flag check within this function, like:

return (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.cares_accept_nodata") &&
            ( status == ARES_ENODATA || status == ARES_ENOTFOUND));

That will make the code outside more readable, like in onAresGetAddrInfoCallback().
This way, you can also avoid adding  accept_nodata_.

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (yanavlasov) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:218]: Can you add a test please, for dual resolutions that if one resolution returns an error and the second is no data, the overall result is no data?

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (wanlill) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc]: Ack, I think I saw that there were also some question on getting rid of accept_nodata_, I wanted to keep it because:

1. To save a call to `Runtime:: runtimeFeatureEnabled`
2. Make sure feature flag won't change while we process the request, I don't think anything would be broken even if this happens but just in case things/code become more complex in the future.

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (wanlill) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc]: sure, added some tests for this.

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (yanjunxiang-google) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:222]: Are we sure the caller always set addrinfo to be nullptr when the status is ARES_ENODATA or ARES_ENOTFOUND?

Others LGTM.

Comment (wanlill) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:222]: I think so, otherwise the current DnsImpl should have been leaking memory in the non-success callback branch, c-ares [doc](https://c-ares.org/ares_getaddrinfo.html) and [code](https://github.com/c-ares/c-ares/blob/main/src/lib/ares_getaddrinfo.c#L399) for reference.

**Comment by yanavlasov** [comment]
I think something got stuck on AZP. Can you add an empty commit or merge main to restart CI please? I will approve and merge after that.

**Comment by yanjunxiang-google** [comment]
LGTM. Defer to @yanavlasov for approval and merge.

**Comment by wanlill** [comment]
@yanavlasov friendly ping, test is green, can you help to take another look?

**Comment by yanavlasov** [comment]
Ok sorry for the delay. One file has merge conflict. Can yo resolve please? You can ping me on Slack and I will enable automerge.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
